### PR TITLE
Added Token Converter Balancer

### PR DESCRIPTION
**Pull request by jpgonzalezra**

### CODE REVIEW

File: contracts/TokenConverterRoute.sol [hunk 1]
```diff
@@ -0,0 +1,64 @@
...
+    }
+    
+    function _getBetterProxy(Token _from, Token _to, uint256 _amount) private view returns (address) {
+        uint minRate = MAX_UINT;
+        address betterProxy = 0x0;
+     
+        for (uint256 i = 0; i < converters.length; i++) {
+            
+            TokenConverter converter = TokenConverter(converters[i]);
+            if (_isAvailable(converter, gasleft())) {
```
Comment (Agusx1211) [inline contracts/TokenConverterRoute.sol]: tx.gasprice should be used: gasLeft() returns the remaining gas limit

File: contracts/TokenConverterRoute.sol [hunk 2]
```diff
@@ -0,0 +1,64 @@
...
+import "./vendors/bancor/converter/BancorGasPriceLimit.sol";
+
+
+contract TokenConverterRoute is TokenConverter, Ownable {
+    
+    uint256 constant internal MAX_UINT = uint256(0) - 1;
+    TokenConverter[] private converters;
+    mapping (address => address) private availability;
+    
+    function addConverter(TokenConverter converter, AvailableProvider availabilityContract) onlyOwner public {
```
Comment (Agusx1211) [inline contracts/TokenConverterRoute.sol]: This setter only allows to add new converters, there should exist the posibility of removing converters and adding/removing availability contracts.

File: contracts/TokenConverterRoute.sol [hunk 3]
```diff
@@ -0,0 +1,64 @@
...
+        converters.push(converter);
+        availability[converter] = availabilityContract;        
+    }
+    
+    function converter(Token _from, Token _to, uint256 _amount, uint256 _minReturn) external payable returns (uint256 amount) {
+        address betterProxy = _getBetterProxy(_from, _to, _amount);
+        TokenConverter converter =  TokenConverter(betterProxy);
+        return converter.convert.value(msg.value)(_from, _to, _amount, _minReturn);
```
Comment (Agusx1211) [inline contracts/TokenConverterRoute.sol]: Tokens should be pulled and then approved for this to work, the converter will try to make a token pull from this contract, and this contract didn't approve that.

File: contracts/TokenConverterRoute.sol [hunk 4]
```diff
@@ -0,0 +1,64 @@
...
+        address betterProxy = _getBetterProxy(_from, _to, _amount);
+        TokenConverter converter =  TokenConverter(betterProxy);
+        return converter.getReturn(_from, _to, _amount);
+    }
+    
+    function _getBetterProxy(Token _from, Token _to, uint256 _amount) private view returns (address) {
+        uint minRate = MAX_UINT;
+        address betterProxy = 0x0;
+     
+        for (uint256 i = 0; i < converters.length; i++) {
```
Comment (Agusx1211) [inline contracts/TokenConverterRoute.sol]: Read converters.length and store it on memory, avoid reading it on each loop

File: contracts/TokenConverterRoute.sol [hunk 5]
```diff
@@ -0,0 +1,64 @@
...
+        return betterProxy;
+    }
+
+    function _isAvailable(address converter, uint256 _gasPrice) private view returns (bool) {
+        
+        if (availability[converter] == 0x0)
+            return AvailableProvider(converter).isAvailable(_gasPrice);            
+            
+        //bancor workaround
+        return (_gasPrice < BancorGasPriceLimit(availability[converter]).gasPrice());
```
Comment (Agusx1211) [inline contracts/TokenConverterRoute.sol]: This can be implemented inside an AvailableProvider

File: contracts/TokenConverterRoute.sol (hunk 2 above)
Comment (jpgonzalezra) [reply contracts/TokenConverterRoute.sol]: Done

File: contracts/TokenConverterRoute.sol (hunk 1 above)
Comment (jpgonzalezra) [reply contracts/TokenConverterRoute.sol]: Done

File: contracts/TokenConverterRoute.sol (hunk 4 above)
Comment (jpgonzalezra) [reply contracts/TokenConverterRoute.sol]: Done

File: contracts/TokenConverterRoute.sol [hunk 6]
```diff
@@ -88,7 +79,7 @@ contract TokenConverterRoute is TokenConverter, Ownable {
             if (_isAvailable(converter, tx.gasprice)) {
                 
                 uint newRate = converter.getReturn(_from, _to, _amount);
-                if  (newRate > 0 && newRate < minRate) {
+                if (newRate > 0 && newRate < minRate) {
```
[the remark this answers is not in the document]
Comment (jpgonzalezra) [reply contracts/TokenConverterRoute.sol]: isTrue ^^

**Pull request merged**